diff --git a/linux/arch/riscv/include/cvm/iie-cvm-sbi.h b/linux/arch/riscv/include/cvm/iie-cvm-sbi.h
index 939cd39c7..feaadcbc1 100644
--- a/linux/arch/riscv/include/cvm/iie-cvm-sbi.h
+++ b/linux/arch/riscv/include/cvm/iie-cvm-sbi.h
@@ -4,6 +4,7 @@
 
 // #include <sbi/sbi_types.h>
 #include <linux/kvm_host.h>
+#include <../uapi/linux/kvm.h>
 
 #define SBI_EXT_CVM				0x20000217
 #define SBI_EXT_CVM_CREATE 					0x0
@@ -13,8 +14,25 @@
 #define SBI_EXT_CVM_FINALIZE				0x4
 #define SBI_EXT_CVM_INIT_MEM_POOL			0x5
 #define SBI_EXT_CVM_RUN_VCPU				0x6
+#define SBI_EXT_CVM_LOAD_FILE				0x7
+#define SBI_EXT_CVM_ENTER					0x8
+#define SBI_EXT_CVM_INIT_PAGE_LIST			0x9
+#define SBI_EXT_CVM_HASH_IMAGE				0xa
+#define SBI_EXT_CVM_ATTEST					0xb
+#define SBI_EXT_CVM_ALLOC_ROOT_PT			0xc
+#define SBI_EXT_CVM_DESTROY					0xd
+#define SBI_EXT_CVM_INIT_SWIOTLB			0xe
+#define SBI_EXT_CVM_REFILL_MEMORY_POOL 		0xf
+#define SBI_EXT_CVM_RETRY_LOAD				0x10
+#define SBI_EXT_RECYCLE_MEMORY				0x11
+/* define for function test */
+#define SBI_EXT_CVM_TEST					0xffff
 
 
+#define PROG_BYK
+#define PROG_LBL
+#define PROG_WSW
+
 struct iie_cvm_vcpu_sbi_params {
 	int *vcpu_id_ptr;	/* id given by userspace at creation */
 	int *vcpu_idx_ptr;	/* index into kvm->vcpu_array */
@@ -27,14 +45,66 @@ struct iie_cvm_sbi_params {
 	int *vcpu_id_ptr;	/* id given by userspace at creation */
 
 	/* G-stage page table */
+	uintptr_t *pgd_phys_ptr;
 	pgd_t *pgd;
-	phys_addr_t *pgd_phys;
+	phys_addr_t pgd_phys;
+	
+	/* SWIOTLB */
+	unsigned long gpa;
+	unsigned long hpa;
 	// struct kvm_vcpu *vcpu;
 	// struct iie_cvm_vcpu_sbi_params cvm_vcpu_sbi_params;
+};
+
+//used for initializing confidential memory
+struct cvm_list_params {
+	unsigned long vaddr;
+	unsigned long addr;
+	unsigned long ele_num;
+	unsigned long page_num;
+	unsigned long level;
+};
 
+//load file physical address array
+struct iie_cvm_sbi_params_load {
+	/* G-stage vmid */
+	struct kvm_vmid *vmid_ptr;
+	unsigned long *src_hpa_array;
+	unsigned long des_gpa;
+	unsigned long count;
 };
 
+struct swiotlb_node {
+	struct swiotlb sw;
+	unsigned long *vmid;
+	struct swiotlb_node *next;
+};
+
+struct cvm_mem_chunk_infor{
+	unsigned long chunk_vaddr;
+	unsigned long chunk_infor_vaddr;
+	unsigned long *paddr_list;
+	unsigned int type;		//initial alloc or realloc
+	bool free;
+	unsigned long *cvm_id;
+};
+
+//use for allocating confidential memory
+#define INITIAL_PAGE_NUM 	13 		//we allocate 2^INITIAL_PAGE_NUM pages for confidential memory at initial time.
+#define REFILL_PAGE_NUM		12		//we allocate 2^REFILL_PAGE_NUM pages to refill the confidential memory pool when it is exhausted.
+#define MAX_CVM_NUM			5		//we have 2^MAX_CVM_NUM confidential virtual machines at most.
+#define CVM_CHUNK_SIZE 		21		//this means the chunk size is 2^21 bytes. Don't modify it.
+#ifndef K
+#define K(x) ((x) << (PAGE_SHIFT-10))
+#endif
+
+#define TEE_NO_MEMORY 	-1
+#define CVM_ERROR		-2
 
 int create_sbi_param(struct kvm *kvm, struct iie_cvm_sbi_params * cvm_sbi_params);
+int cvm_mem_manege_init(void);
+unsigned long iie_kernel_page_translate(unsigned long addr);
+int refill_KVM_memory_pool(void);
+//int kvm_vm_ioctl_load_file(struct kvm *kvm, void *argp);
 
 #endif
\ No newline at end of file
diff --git a/linux/arch/riscv/kvm/vcpu.c b/linux/arch/riscv/kvm/vcpu.c
index ada386ebe..8ebc2e10c 100644
--- a/linux/arch/riscv/kvm/vcpu.c
+++ b/linux/arch/riscv/kvm/vcpu.c
@@ -21,7 +21,10 @@
 #include <asm/cacheflush.h>
 #include <asm/kvm_vcpu_vector.h>
 #include <cvm/iie-cvm-sbi.h>
-
+#include <asm/pgtable-bits.h>
+#include <asm/page.h>
+#include <linux/of.h>
+#include <linux/uaccess.h>
 
 const struct _kvm_stats_desc kvm_vcpu_stats_desc[] = {
 	KVM_GENERIC_VCPU_STATS(),
@@ -146,20 +149,21 @@ int kvm_arch_vcpu_create(struct kvm_vcpu *vcpu)
 	/* Reset VCPU */
 	kvm_riscv_reset_vcpu(vcpu);
 
-	struct iie_cvm_sbi_params *cvm_sbi_params = kmalloc(sizeof(struct iie_cvm_sbi_params), GFP_KERNEL);
-	cvm_sbi_params->vmid_ptr = __pa(&vcpu->kvm->arch.vmid);
-	cvm_sbi_params->vcpu_id_ptr = __pa(&vcpu->vcpu_id);
-	cvm_sbi_params->pgd = __pa(vcpu->kvm->arch.pgd);
-	cvm_sbi_params->pgd_phys = __pa(&vcpu->kvm->arch.pgd_phys);
-	
+	if(vcpu->kvm->cmode)
+	{
+		// #ifdef PROG_WSW
 
-	uintptr_t pa_cvm = __pa(cvm_sbi_params);
-	// printk("pa_cvm = %ld\n", pa_cvm);
-	sbi_ecall(SBI_EXT_CVM, SBI_EXT_CVM_CREATE,
-		  pa_cvm, 0, 0, 0, 0, 0);
-	struct sbiret retval = sbi_ecall(SBI_EXT_CVM, SBI_EXT_CVM_CREATE_VCPU,
-		  pa_cvm, 0, 0, 0, 0, 0);
+		struct iie_cvm_sbi_params *cvm_sbi_params = kmalloc(sizeof(struct iie_cvm_sbi_params), GFP_KERNEL);
+		cvm_sbi_params->vmid_ptr = __pa(&vcpu->kvm->arch.vmid);
+		cvm_sbi_params->vcpu_id_ptr = __pa(&vcpu->vcpu_id);
+		uintptr_t pa_cvm = __pa(cvm_sbi_params);
 
+		struct sbiret retval = sbi_ecall(SBI_EXT_CVM, SBI_EXT_CVM_CREATE_VCPU, pa_cvm, 0, 0, 0, 0, 0);
+		//kfree(cvm_sbi_params);
+		// #endif
+	}
+
+	
 
 	return 0;
 }
@@ -638,31 +642,28 @@ static void noinstr kvm_riscv_vcpu_enter_exit(struct kvm_vcpu *vcpu)
 	guest_state_exit_irqoff();
 }
 
+
+struct swiotlb_node *sw_list_head;
+static DEFINE_SPINLOCK(sw_list);
+
+
 int kvm_arch_vcpu_ioctl_run(struct kvm_vcpu *vcpu)
 {
 	int ret;
-	struct kvm_cpu_trap trap;
+	struct sbiret sbi_ret;
+	struct kvm_cpu_trap *trap = kmalloc(sizeof(struct kvm_cpu_trap), GFP_KERNEL);
 	struct kvm_run *run = vcpu->run;
-
 	struct iie_cvm_sbi_params *cvm_sbi_params = kmalloc(sizeof(struct iie_cvm_sbi_params), GFP_KERNEL);
+
+	sbi_ret.error = 0;
 	cvm_sbi_params->vmid_ptr = __pa(&vcpu->kvm->arch.vmid);
 	cvm_sbi_params->vcpu_id_ptr = __pa(&vcpu->vcpu_id);
-	cvm_sbi_params->pgd = __pa(vcpu->kvm->arch.pgd);
-	cvm_sbi_params->pgd_phys = __pa(&vcpu->kvm->arch.pgd_phys);
-	uintptr_t pa_cvm = __pa(cvm_sbi_params);
-
-	if (!vcpu->arch.ran_atleast_once)
-	{		
-		struct sbiret retval = sbi_ecall(SBI_EXT_CVM, SBI_EXT_CVM_RUN_VCPU,
-		  pa_cvm, 0, 0, 0, 0, 0);
-		// printk("Kernel: RUN retval.value = %d\tretval.error = %d\n", retval.value, retval.error);
-	}
 
 	/* Mark this VCPU ran at least once */
 	vcpu->arch.ran_atleast_once = true;
 
 	kvm_vcpu_srcu_read_lock(vcpu);
-
+	
 	switch (run->exit_reason) {
 	case KVM_EXIT_MMIO:
 		/* Process MMIO value returned from user-space */
@@ -693,7 +694,6 @@ int kvm_arch_vcpu_ioctl_run(struct kvm_vcpu *vcpu)
 	vcpu_load(vcpu);
 
 	kvm_sigset_activate(vcpu);
-
 	ret = 1;
 	run->exit_reason = KVM_EXIT_UNKNOWN;
 	while (ret > 0) {
@@ -715,7 +715,6 @@ int kvm_arch_vcpu_ioctl_run(struct kvm_vcpu *vcpu)
 			preempt_enable();
 			continue;
 		}
-
 		local_irq_disable();
 
 		/*
@@ -759,7 +758,43 @@ int kvm_arch_vcpu_ioctl_run(struct kvm_vcpu *vcpu)
 
 		guest_timing_enter_irqoff();
 
-		kvm_riscv_vcpu_enter_exit(vcpu);
+		if(vcpu->kvm->cmode)
+		{
+			// #ifdef PROG_LBL
+			struct swiotlb_node *node;
+			spin_lock(&sw_list);
+			for(struct swiotlb_node *cur = sw_list_head; cur->next; cur = cur->next){
+                node = cur->next;
+                if(node->vmid == &vcpu->kvm->arch.vmid.vmid){
+                break;
+            	}
+            }
+			spin_unlock(&sw_list);
+			cvm_sbi_params->gpa = (trap->htval << 2) | (trap->stval & 0x3);
+			if(cvm_sbi_params->gpa >= node->sw.addr && cvm_sbi_params->gpa < node->sw.addr + node->sw.size){
+				pte_t *ptep;
+				u32 ptep_level;
+				bool found_leaf;
+				found_leaf = iie_gstage_get_leaf_entry(vcpu->kvm, cvm_sbi_params->gpa, &ptep, &ptep_level);
+				if(found_leaf){
+					cvm_sbi_params->hpa = (unsigned long)(pte_val(*ptep) >> 10 << PAGE_SHIFT);
+					//printk("cvm_sbi_params hpa is 0x%lx\n", cvm_sbi_params->hpa);
+				}
+			}
+			// #endif
+
+			// #ifdef PROG_BYK
+			sbi_ret = sbi_ecall(SBI_EXT_CVM, SBI_EXT_CVM_RUN_VCPU, __pa(cvm_sbi_params),
+				__pa(&vcpu->arch.guest_context), __pa(&vcpu->arch.guest_csr), __pa(trap), 0, 0);
+			// #endif
+		}
+		
+		if(!vcpu->kvm->cmode)
+		{
+			// #ifndef PROG_BYK
+			kvm_riscv_vcpu_enter_exit(vcpu);
+			// #endif
+		}
 
 		vcpu->mode = OUTSIDE_GUEST_MODE;
 		vcpu->stat.exits++;
@@ -769,11 +804,16 @@ int kvm_arch_vcpu_ioctl_run(struct kvm_vcpu *vcpu)
 		 * get an interrupt between __kvm_riscv_switch_to() and
 		 * local_irq_enable() which can potentially change CSRs.
 		 */
-		trap.sepc = vcpu->arch.guest_context.sepc;
-		trap.scause = csr_read(CSR_SCAUSE);
-		trap.stval = csr_read(CSR_STVAL);
-		trap.htval = csr_read(CSR_HTVAL);
-		trap.htinst = csr_read(CSR_HTINST);
+		if(!vcpu->kvm->cmode)
+		{
+			// #ifndef PROG_BYK
+			trap->sepc = vcpu->arch.guest_context.sepc;
+			trap->scause = csr_read(CSR_SCAUSE);
+			trap->stval = csr_read(CSR_STVAL);
+			trap->htval = csr_read(CSR_HTVAL);
+			trap->htinst = csr_read(CSR_HTINST);
+			// #endif
+		}
 
 		/* Syncup interrupts state with HW */
 		kvm_riscv_vcpu_sync_interrupts(vcpu);
@@ -789,6 +829,7 @@ int kvm_arch_vcpu_ioctl_run(struct kvm_vcpu *vcpu)
 		 * interrupts between the enable and disable.
 		 */
 		local_irq_enable();
+
 		local_irq_disable();
 
 		guest_timing_exit_irqoff();
@@ -796,10 +837,19 @@ int kvm_arch_vcpu_ioctl_run(struct kvm_vcpu *vcpu)
 		local_irq_enable();
 
 		preempt_enable();
+		
+		if(sbi_ret.error == TEE_NO_MEMORY){
+			// printk("refill_KVM_memory_pool function here!\n");
+			ret = refill_KVM_memory_pool();
+			if (ret)
+				return ret;
+			/* If we got host interrupt then do nothing */
+			trap->scause = CAUSE_IRQ_FLAG;
+		}
 
 		kvm_vcpu_srcu_read_lock(vcpu);
 
-		ret = kvm_riscv_vcpu_exit(vcpu, run, &trap);
+		ret = kvm_riscv_vcpu_exit(vcpu, run, trap);
 	}
 
 	kvm_sigset_deactivate(vcpu);
@@ -808,5 +858,48 @@ int kvm_arch_vcpu_ioctl_run(struct kvm_vcpu *vcpu)
 
 	kvm_vcpu_srcu_read_unlock(vcpu);
 
+	kfree(cvm_sbi_params);
 	return ret;
 }
+
+int kvm_vm_ioctl_swiotlb(struct kvm *kvm, void *argp){
+	if(!sw_list_head){
+		sw_list_head = kmalloc(sizeof(struct swiotlb_node), GFP_KERNEL);
+		sw_list_head->next = NULL;
+	}
+	struct swiotlb_node *sw_node = kmalloc(sizeof(struct swiotlb_node), GFP_KERNEL);
+	if (copy_from_user(&sw_node->sw, argp, sizeof(struct swiotlb))){
+		printk("kvm_vm_ioctl_swiotlb copy iotcl parameters failed!\n");
+        return -1;
+    }
+	sw_node->vmid = &kvm->arch.vmid.vmid;
+	spin_lock(&sw_list);
+	sw_node->next = sw_list_head->next;
+	sw_list_head->next = sw_node;
+	spin_unlock(&sw_list);
+	sbi_ecall(SBI_EXT_CVM, SBI_EXT_CVM_INIT_SWIOTLB, __pa(&sw_node->sw), __pa(&kvm->arch.vmid), 0, 0, 0, 0);
+	return 0;
+}
+
+int kvm_riscv_destroy_sw_node(struct kvm *kvm){
+	if(!sw_list_head){
+		return 0;
+	}
+	struct swiotlb_node *node, *cur;
+	spin_lock(&sw_list);
+	for(cur = sw_list_head; cur->next; cur = cur->next){
+		node = cur->next;
+		// sbi_printf("[IIE CVM Monitor@%s] node vmid id = %d\tvmid = %d\n", __func__, node->cvm.vmid->vmid, vmid);
+		if(node->vmid == &kvm->arch.vmid.vmid){
+			cur->next = node->next;
+			spin_unlock(&sw_list);
+			//printk("HOST OS DESTROY CVM SW NODE!\n");
+			kfree(node);
+			return 0;
+		}
+	}
+	spin_unlock(&sw_list);
+	return 1;
+}
+
+
diff --git a/linux/arch/riscv/kvm/vm.c b/linux/arch/riscv/kvm/vm.c
index 7e2b50c69..6e5ee5a07 100644
--- a/linux/arch/riscv/kvm/vm.c
+++ b/linux/arch/riscv/kvm/vm.c
@@ -11,6 +11,7 @@
 #include <linux/module.h>
 #include <linux/uaccess.h>
 #include <linux/kvm_host.h>
+#include <cvm/iie-cvm-sbi.h>
 
 const struct _kvm_stats_desc kvm_vm_stats_desc[] = {
 	KVM_GENERIC_VM_STATS()
@@ -50,6 +51,37 @@ int kvm_arch_init_vm(struct kvm *kvm, unsigned long type)
 
 void kvm_arch_destroy_vm(struct kvm *kvm)
 {
+	unsigned long i;
+	if(kvm->cmode)
+	{
+		// #ifdef PROG_WSW
+		struct iie_cvm_sbi_params *cvm_sbi_params = kmalloc(sizeof(struct iie_cvm_sbi_params), GFP_KERNEL);
+		cvm_sbi_params->vmid_ptr = __pa(&kvm->arch.vmid);
+		uintptr_t pa_cvm = __pa(cvm_sbi_params);
+
+		////#ifdef PROG_LBL
+		struct cvm_list_params *ret_params = kzalloc(sizeof(struct cvm_list_params), GFP_KERNEL);
+		unsigned long *chunk_vaddr_list = (unsigned long *)__get_free_pages(GFP_KERNEL, get_order(2^10*PAGE_SIZE));
+		while(!chunk_vaddr_list){
+			chunk_vaddr_list = (unsigned long *)__get_free_pages(GFP_KERNEL, get_order(2^10*PAGE_SIZE));
+		}
+		ret_params->addr = (unsigned long)__pa(chunk_vaddr_list);
+		uintptr_t pa_ret = __pa(ret_params);
+		sbi_ecall(SBI_EXT_CVM, SBI_EXT_RECYCLE_MEMORY, pa_cvm, __pa(ret_params), 0, 0, 0, 0);
+		struct cvm_mem_chunk_infor *chunk_infor;
+		// printk("recycle %ld chunks\n", ret_params->ele_num);
+		for(i=0; i<ret_params->ele_num; i++){
+			chunk_infor = (struct cvm_mem_chunk_infor *)*((unsigned long *)__va(ret_params->addr)+i);
+			vfree((void *)chunk_infor->chunk_vaddr);
+			__free_pages(virt_to_page(chunk_infor->chunk_infor_vaddr), 0);
+		}
+
+		sbi_ecall(SBI_EXT_CVM, SBI_EXT_CVM_DESTROY, pa_cvm, pa_ret, 0, 0, 0, 0);
+		// #endif
+		kvm_riscv_destroy_sw_node(kvm);
+		//#endif
+		kfree(cvm_sbi_params);
+	}
 	kvm_destroy_vcpus(kvm);
 
 	kvm_riscv_aia_destroy_vm(kvm);
diff --git a/linux/include/linux/kvm_host.h b/linux/include/linux/kvm_host.h
index fb6c6109f..84f51c299 100644
--- a/linux/include/linux/kvm_host.h
+++ b/linux/include/linux/kvm_host.h
@@ -804,6 +804,8 @@ struct kvm {
 	bool vm_bugged;
 	bool vm_dead;
 
+	bool cmode;
+
 #ifdef CONFIG_HAVE_KVM_PM_NOTIFIER
 	struct notifier_block pm_notifier;
 #endif
@@ -1444,7 +1446,8 @@ int kvm_arch_vcpu_ioctl_set_mpstate(struct kvm_vcpu *vcpu,
 int kvm_arch_vcpu_ioctl_set_guest_debug(struct kvm_vcpu *vcpu,
 					struct kvm_guest_debug *dbg);
 int kvm_arch_vcpu_ioctl_run(struct kvm_vcpu *vcpu);
-
+int kvm_vm_ioctl_swiotlb(struct kvm *kvm, void *argp);
+int kvm_riscv_destroy_sw_node(struct kvm *kvm);
 void kvm_arch_sched_in(struct kvm_vcpu *vcpu, int cpu);
 
 void kvm_arch_vcpu_load(struct kvm_vcpu *vcpu, int cpu);
diff --git a/linux/include/uapi/linux/kvm.h b/linux/include/uapi/linux/kvm.h
index 13065dd96..25266614c 100644
--- a/linux/include/uapi/linux/kvm.h
+++ b/linux/include/uapi/linux/kvm.h
@@ -1474,6 +1474,17 @@ struct kvm_vfio_spapr_tce {
 #define KVM_SET_TSS_ADDR          _IO(KVMIO,   0x47)
 #define KVM_SET_IDENTITY_MAP_ADDR _IOW(KVMIO,  0x48, __u64)
 
+struct load_file{
+	unsigned long src_hva;
+	unsigned long des_gpa;
+	unsigned long file_size;
+};
+struct swiotlb{
+	unsigned long addr;
+	unsigned long size;
+};
+#define KVM_LOAD_FILE		  	_IOWR(KVMIO, 0x49, struct load_file)
+#define KVM_SET_SWIOTLB		  	_IOWR(KVMIO, 0x53, struct swiotlb)
 /* enable ucontrol for s390 */
 struct kvm_s390_ucas_mapping {
 	__u64 user_addr;
diff --git a/linux/virt/kvm/kvm_main.c b/linux/virt/kvm/kvm_main.c
index 486800a70..e4af6e77c 100644
--- a/linux/virt/kvm/kvm_main.c
+++ b/linux/virt/kvm/kvm_main.c
@@ -61,7 +61,7 @@
 #include "async_pf.h"
 #include "kvm_mm.h"
 #include "vfio.h"
-
+#include <linux/memblock.h>
 #include <trace/events/ipi.h>
 
 #define CREATE_TRACE_POINTS
@@ -69,6 +69,7 @@
 
 #include <linux/kvm_dirty_ring.h>
 
+#include <cvm/iie-cvm-sbi.h>
 
 /* Worst case buffer size needed for holding an integer. */
 #define ITOA_MAX_LEN 12
@@ -1163,6 +1164,7 @@ static struct kvm *kvm_create_vm(unsigned long type, const char *fdname)
 	KVM_MMU_LOCK_INIT(kvm);
 	mmgrab(current->mm);
 	kvm->mm = current->mm;
+	kvm->cmode = type == SBI_EXT_CVM ? 1 : 0;
 	kvm_eventfd_init(kvm);
 	mutex_init(&kvm->lock);
 	mutex_init(&kvm->irq_lock);
@@ -1254,6 +1256,33 @@ static struct kvm *kvm_create_vm(unsigned long type, const char *fdname)
 	preempt_notifier_inc();
 	kvm_init_pm_notifier(kvm);
 
+	if(kvm->cmode) 
+	{
+		printk("------ %s: cmode activate. --------\n", __func__);
+		static bool cvm_mem_inited = false;
+		if(!cvm_mem_inited)
+		{
+			// #ifdef PROG_LBL
+			r = cvm_mem_manege_init();
+			if(r)
+				goto out_err_no_disable;
+			cvm_mem_inited = true;
+			// #endif
+		}
+		struct iie_cvm_sbi_params *cvm_sbi_params = kmalloc(sizeof(struct iie_cvm_sbi_params), GFP_KERNEL);
+		cvm_sbi_params->vmid_ptr = __pa(&kvm->arch.vmid);
+		uintptr_t pa_cvm = __pa(cvm_sbi_params);
+
+		// #ifdef PROG_WSW
+		sbi_ecall(SBI_EXT_CVM, SBI_EXT_CVM_CREATE, pa_cvm, 0, 0, 0, 0, 0);
+		// #endif
+
+		// #ifdef PROG_LBL
+		sbi_ecall(SBI_EXT_CVM, SBI_EXT_CVM_ALLOC_ROOT_PT, pa_cvm, 0, 0, 0, 0, 0);
+		kfree(cvm_sbi_params);
+		// #endif
+	}
+
 	return kvm;
 
 out_err:
@@ -4030,6 +4059,25 @@ static int kvm_vm_ioctl_create_vcpu(struct kvm *kvm, u32 id)
 	mutex_unlock(&kvm->lock);
 	kvm_arch_vcpu_postcreate(vcpu);
 	kvm_create_vcpu_debugfs(vcpu);
+
+	if(vcpu->kvm->cmode)
+	{
+		// #ifdef PROG_WSW
+
+		struct iie_cvm_sbi_params *cvm_sbi_params = kmalloc(sizeof(struct iie_cvm_sbi_params), GFP_KERNEL);
+		cvm_sbi_params->vmid_ptr = __pa(&vcpu->kvm->arch.vmid);
+		cvm_sbi_params->vcpu_id_ptr = __pa(&vcpu->vcpu_id);
+		cvm_sbi_params->pgd = __pa(vcpu->kvm->arch.pgd);
+		cvm_sbi_params->pgd_phys = __pa(&vcpu->kvm->arch.pgd_phys);
+		uintptr_t pa_cvm = __pa(cvm_sbi_params);
+
+		if(vcpu->vcpu_id == 1){
+			sbi_ecall(SBI_EXT_CVM, SBI_EXT_CVM_TEST, pa_cvm, 0, 0, 0, 0, 0);
+		}
+		// #endif
+		kfree(cvm_sbi_params);
+	}
+
 	return r;
 
 kvm_put_xa_release:
@@ -4788,6 +4836,68 @@ static int kvm_vm_ioctl_get_stats_fd(struct kvm *kvm)
 	return fd;
 }
 
+static int kvm_vm_ioctl_load_file(struct kvm *kvm, void *argp){
+	// printk("----------------linux vm ioctl begin to load file-------------------\n");
+	struct sbiret ret;
+	unsigned long file_paddr_array;
+	struct load_file load_file_t;
+	unsigned long va;
+	int r = -ENOMEM;
+	if (copy_from_user(&load_file_t, argp, sizeof(struct load_file))){
+		printk("Copy iotcl parameters failed!\n");
+		return -1;
+	}
+	unsigned long pages_num = (load_file_t.file_size + PAGE_SIZE - 1) / PAGE_SIZE;
+	unsigned long addr_entry_num = (pages_num + 512 - 1) / 512;
+	unsigned long virt_address = __get_free_pages(GFP_KERNEL, get_order(addr_entry_num * PAGE_SIZE));
+	unsigned long *src_hpa_array = (unsigned long *)virt_address;
+	
+	// unsigned long hva;
+	// unsigned long hpa;
+	// for(int i=0;i<pages_num;i++){
+	// 	hpa = virt_to_phys((void *)(load_file_t.src_hva+i*PAGE_SIZE));
+	// 	*(src_hpa_array+i) = hpa;
+	// 	for(int j=0;j<512;j++){
+	// 		printk("%lx\n", *((unsigned long *)hpa + j));
+	// 	}
+	// }
+	for(int i=0;i<pages_num;i++){
+		va = __get_free_pages(GFP_KERNEL, 0);
+		if(copy_from_user((void *)va, (void *)(load_file_t.src_hva+i*PAGE_SIZE), PAGE_SIZE)){
+			printk("copy_from_user failed!\n");
+			return 0;
+		}else{
+			*(src_hpa_array+i) = __pa(va);
+		}
+	}
+
+	file_paddr_array = __pa(virt_address);
+	unsigned long va_sbi_params = __get_free_pages(GFP_KERNEL, 0);
+	struct iie_cvm_sbi_params_load *sbi_load_params = (struct iie_cvm_sbi_params_load *)va_sbi_params;
+	sbi_load_params->des_gpa = load_file_t.des_gpa;
+	sbi_load_params->count = pages_num;
+	sbi_load_params->src_hpa_array = (unsigned long*)file_paddr_array;
+	sbi_load_params->vmid_ptr = (struct kvm_vmid *)__pa(&kvm->arch.vmid);
+	ret = sbi_ecall(SBI_EXT_CVM, SBI_EXT_CVM_LOAD_FILE, __pa(va_sbi_params), 0, 0, 0, 0, 0);
+	while(ret.error == TEE_NO_MEMORY){
+		// printk("refill_KVM_memory_pool here\n");
+		r = refill_KVM_memory_pool();
+		if(r)
+			return r;
+		ret = sbi_ecall(SBI_EXT_CVM, SBI_EXT_CVM_RETRY_LOAD, __pa(va_sbi_params), 0, 0, 0, 0, 0);
+	}
+
+
+	for(int i=0;i<pages_num;i++){
+		va = (unsigned long)__va(*(src_hpa_array+i));
+		__free_pages(virt_to_page(va), 0);
+	}
+	__free_pages(virt_to_page(virt_address), get_order(addr_entry_num * PAGE_SIZE));
+	__free_pages(virt_to_page(va_sbi_params), 0);
+	// printk("----------------linux vm ioctl end to load file-------------------\n");
+	return ret.error;
+}
+
 static long kvm_vm_ioctl(struct file *filp,
 			   unsigned int ioctl, unsigned long arg)
 {
@@ -4973,6 +5083,12 @@ static long kvm_vm_ioctl(struct file *filp,
 	case KVM_GET_STATS_FD:
 		r = kvm_vm_ioctl_get_stats_fd(kvm);
 		break;
+	case KVM_LOAD_FILE:
+		r = kvm_vm_ioctl_load_file(kvm, argp);
+		break;
+	case KVM_SET_SWIOTLB:
+		r = kvm_vm_ioctl_swiotlb(kvm, argp);
+		break;
 	default:
 		r = kvm_arch_vm_ioctl(filp, ioctl, arg);
 	}
@@ -6264,3 +6380,237 @@ int kvm_vm_create_worker_thread(struct kvm *kvm, kvm_vm_thread_fn_t thread_fn,
 
 	return init_context.err;
 }
+
+// used for translate address allocated by vmalloc.
+unsigned long iie_kernel_page_translate(unsigned long addr){
+	pgd_t *pgd;
+	pud_t *pud;
+	p4d_t *p4d;
+	pmd_t *pmd;
+	pte_t *pte;
+
+	pgd = pgd_offset_k(addr);
+	if (!pgd_present(*pgd))
+		return false;
+	if (pgd_leaf(*pgd))
+		return true;
+
+	p4d = p4d_offset(pgd, addr);
+	if (!p4d_present(*p4d))
+		return false;
+	if (p4d_leaf(*p4d))
+		return true;
+
+	pud = pud_offset(p4d, addr);
+	if (!pud_present(*pud))
+		return false;
+	if (pud_leaf(*pud))
+		return true;
+
+	pmd = pmd_offset(pud, addr);
+	if (!pmd_present(*pmd))
+		return false;
+	if (pmd_leaf(*pmd))
+		return true;
+
+	pte = pte_offset_kernel(pmd, addr);
+	return (pte_val(*pte) >> _PAGE_PFN_SHIFT);
+}
+
+int cvm_init_bitmap(struct cvm_list_params *bitmap){
+	unsigned long host_page_num, bitmap_page_num;
+	unsigned long *bitmap_addr;
+
+	host_page_num = get_num_physpages();
+	bitmap_page_num = (host_page_num / 8 + PAGE_SIZE -1) / PAGE_SIZE;
+	bitmap->page_num = bitmap_page_num;
+	bitmap_addr = (unsigned long *)__get_free_pages(GFP_KERNEL, get_order(host_page_num / 8));			//used for bitmap
+	if(!bitmap_addr){
+		printk("bitmap alloc failed!\n");
+		return -ENOMEM;
+	}
+	bitmap->addr = __pa((unsigned long)bitmap_addr);
+	//bitmap element is unsigned long type.
+	bitmap->ele_num = host_page_num / 64;
+	return 0;
+}
+
+int cvm_init_ownershiptable(struct cvm_list_params *page_own_table){
+	unsigned long host_page_num, page_own_table_num, level1, i, page_vaddr;
+	unsigned long *page_own_table_addr;
+	
+	host_page_num = get_num_physpages();
+	page_own_table_num = (host_page_num * 8 + PAGE_SIZE -1) / PAGE_SIZE;
+	page_own_table->page_num = page_own_table_num;
+	if(page_own_table_num <= 1024){
+		level1=0;
+		page_own_table_addr = (unsigned long *)__get_free_pages(GFP_KERNEL, get_order(host_page_num * 8));
+		if(!page_own_table_addr){
+			printk("page_own_table alloc failed when page_own_table_num is less and equal than 2^10!\n");
+			return -ENOMEM;
+		}
+	}else{
+		level1 = (page_own_table_num + 1024 - 1) / 1024;
+		if(level1 <= 512){
+			// printk("page_own_table_num is %ld, and level1 is %ld\n", page_own_table_num, level1);
+			//we don't need level2 Ownership page table;
+			page_own_table_addr = (unsigned long *)__get_free_pages(GFP_KERNEL, 0);
+			if(!page_own_table_addr){
+				printk("page_own_table alloc failed when page_own_table_num is greater than 2^10!\n");
+				return -ENOMEM;
+			}
+			for(i=0; i < level1; i++){
+				page_vaddr = (unsigned long)__get_free_pages(GFP_KERNEL, get_order(1024 * PAGE_SIZE));  //allocate 4MB memory for ownership page table.
+				if(!page_vaddr){
+					printk("page_vaddr for Ownership page table alloc failed!\n");
+					return -ENOMEM;
+				}
+				//  printk("%dth page paddr is %lx, vaddr is %lx\n", i, __pa(page_vaddr), page_vaddr);
+				*((unsigned long *)page_own_table_addr + i) = __pa(page_vaddr);
+			}
+		}else{
+			printk("host memory exceed 1TB!\n");
+			return -ENOMEM;
+		}
+	}
+	page_own_table->level = level1;
+	page_own_table->addr = __pa((unsigned long)page_own_table_addr);
+	//bitmap element is unsigned long type.
+	page_own_table->ele_num = host_page_num;
+	return 0;
+}
+
+int cvm_init_rootpagetable(struct cvm_list_params *root_pt_params, unsigned long *rpt_addr_list){
+	unsigned long root_pt_va, i;
+	//we already known the max number of cvm is 32
+	// unsigned long *root_pt_list = (unsigned long *)__get_free_pages(GFP_KERNEL, 0);
+	for(i=0; i<(1<<MAX_CVM_NUM); i++){
+		root_pt_va = __get_free_pages(GFP_KERNEL, 2);
+		if(!root_pt_va){
+			printk("root_pt_va alloc failed!\n");
+			return -ENOMEM;
+		}
+		*(rpt_addr_list + i) = __pa(root_pt_va);
+		
+	}
+	root_pt_params->addr = __pa((unsigned long)rpt_addr_list);
+	root_pt_params->ele_num = (1UL<<MAX_CVM_NUM);
+	return 0;
+}
+
+int cvm_init_memorypool(struct cvm_list_params *chunk_infor_list, unsigned long *chunk_infor_addr, unsigned long chunk_num, bool isrefill){
+	unsigned long *chunk_vaddr, i, j;
+	chunk_infor_list->ele_num = chunk_num;
+	chunk_infor_list->page_num = chunk_num >> 9;
+	for(i=0; i<chunk_num; i++){
+		struct cvm_mem_chunk_infor *chunk_infor = (struct cvm_mem_chunk_infor *)__get_free_pages(GFP_KERNEL, 0);
+		if (!chunk_infor){
+			printk("chunnk_info vmalloc failed!\n");
+			return -ENOMEM;
+		}
+		chunk_vaddr = vmalloc(PAGE_SIZE * 512);
+		if (!chunk_vaddr){
+			printk("chunk_vaddr vmalloc failed!\n");
+			return -ENOMEM;
+		}
+		unsigned long *chunk_paddr_list = (unsigned long *)__get_free_pages(GFP_KERNEL, 0);
+		if(!chunk_paddr_list){
+			printk("chunk_paddr_list malloc failed!\n");
+			return -ENOMEM;
+		}
+		for(j=0; j<512; j++){
+			*(chunk_paddr_list+j) = iie_kernel_page_translate((unsigned long)chunk_vaddr + j*PAGE_SIZE) << PAGE_SHIFT;
+			// printk("%lx\n", *(chunk_paddr_list+j));
+		}
+		
+		chunk_infor->chunk_vaddr = (unsigned long)chunk_vaddr;
+		chunk_infor->paddr_list = (unsigned long *)__pa(chunk_paddr_list);
+		*(chunk_infor_addr + i) = __pa(chunk_infor);
+		chunk_infor->chunk_infor_vaddr = (unsigned long)chunk_infor;
+		if(isrefill)
+			chunk_infor->type = 1; 	//They need to be recycled when the CVMs are destroyed.
+		else
+			chunk_infor->type = 0;	//They don't need to be recycled.
+	}
+	chunk_infor_list->addr = __pa(chunk_infor_addr);
+
+
+	return 0;
+}
+
+int cvm_mem_manege_init(){
+	struct cvm_list_params *bitmap, *page_own_table, *root_pt, *chunk_infor_list;
+	unsigned long *rpt_addr_list, *chunk_infor_addr;
+	unsigned long chunk_num;
+	int re;
+	struct sbiret ret;
+
+	bitmap = (struct cvm_list_params *)__get_free_pages(GFP_KERNEL, 0);
+	re = cvm_init_bitmap(bitmap);
+	if(re != 0){
+		return re;
+	}
+
+	page_own_table = (struct cvm_list_params *)__get_free_pages(GFP_KERNEL, 0);
+	re = cvm_init_ownershiptable(page_own_table);
+	if(re != 0){
+		return re;
+	}
+
+	rpt_addr_list = (unsigned long *)__get_free_pages(GFP_KERNEL, 0);
+	root_pt = (struct cvm_list_params *)__get_free_pages(GFP_KERNEL, 0);
+	re = cvm_init_rootpagetable(root_pt, rpt_addr_list);
+	if(re != 0){
+		return re;
+	}
+	//the size of chunk is 2MB.Because one page can only contain 512 addresses of physical page while 512 physical pages are 2MB.
+	chunk_num = (PAGE_SIZE  << INITIAL_PAGE_NUM) >> CVM_CHUNK_SIZE; 
+	chunk_infor_list = kmalloc(sizeof(struct cvm_list_params), GFP_KERNEL | __GFP_ZERO);
+	chunk_infor_addr = (unsigned long *)__get_free_pages(GFP_KERNEL, 0);
+	re = cvm_init_memorypool(chunk_infor_list, chunk_infor_addr, chunk_num, false);
+	if(re != 0){
+		return re;
+	}
+
+	ret = sbi_ecall(SBI_EXT_CVM, SBI_EXT_CVM_INIT_PAGE_LIST, __pa((unsigned long)chunk_infor_list), 
+		__pa((unsigned long)root_pt), __pa((unsigned long)bitmap), __pa((unsigned long)page_own_table), 0, 0);
+
+	__free_pages(virt_to_page(chunk_infor_addr), 0);
+	__free_pages(virt_to_page(root_pt), 0);
+	__free_pages(virt_to_page(bitmap), 0);
+	__free_pages(virt_to_page(page_own_table), 0);
+	// __free_pages(virt_to_page(cm_addr_list), get_order(cm_list_page_num * PAGE_SIZE));
+	__free_pages(virt_to_page(rpt_addr_list), 0);
+
+	return ret.error;
+}
+
+int refill_KVM_memory_pool(){
+	// printk("Host OS refilled confidential memory pool!\n");
+	struct sbiret ret;
+	unsigned long *chunk_infor_addr;
+	unsigned long refill_chunk_num;
+	int re;
+	struct cvm_list_params *refill_chunk_list;
+
+	refill_chunk_num = PAGE_SIZE << REFILL_PAGE_NUM >> CVM_CHUNK_SIZE;
+	// printk("%d\n",refill_chunk_num);
+	refill_chunk_list = (struct cvm_list_params *)__get_free_pages(GFP_KERNEL, 0);
+	if(!refill_chunk_list){
+		printk("Failed to allocate refill_chunk_list!\n");
+		return -ENOMEM;
+	}
+	chunk_infor_addr = (unsigned long *)__get_free_pages(GFP_KERNEL, 0);
+	if(!chunk_infor_addr){
+		printk("Failed to allocate chunk_infor_addr!\n");
+		return -ENOMEM;
+	}
+	re = cvm_init_memorypool(refill_chunk_list, chunk_infor_addr, refill_chunk_num, true);
+	if(re != 0){
+		return re;
+	}
+	ret = sbi_ecall(SBI_EXT_CVM, SBI_EXT_CVM_REFILL_MEMORY_POOL, __pa(refill_chunk_list), 0, 0, 0, 0, 0);
+	__free_pages(virt_to_page(chunk_infor_addr), 0);
+	// printk("Host OS end refilled confidential memory pool!\n");
+	return ret.error;
+}
\ No newline at end of file
